official/recommendation/neumf_model.py

```diff
@@ -0,0 +1,130 @@
+# Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Defines NeuMF model for NCF framework.
+
+Some abbreviations used in the code base:
+NeuMF: Neural Matrix Factorization
+NCF: Neural Collaborative Filtering
+GMF: Generalized Matrix Factorization
+MLP: Multi-Layer Perceptron
+
+GMF applies a linear kernel to model the latent feature interactions, and MLP
+uses a nonlinear kernel to learn the interaction function from data. NeuMF model
+is a fused model of GMF and MLP to better model the complex user-item
+interactions, and unifies the strengths of linearity of MF and non-linearity of
+MLP for modeling the user-item latent structures.
+
+In NeuMF model, it allows GMF and MLP to learn separate embeddings, and combine
+the two models by concatenating their last hidden layer.
+"""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from six.moves import xrange  # pylint: disable=redefined-builtin
+import tensorflow as tf
+
+from official.recommendation import constants  # pylint: disable=g-bad-import-order
+
+
+class NeuMF(tf.keras.models.Model):
+  """Neural matrix factorization (NeuMF) model for recommendations."""
+
+  def __init__(self, num_users, num_items, mf_dim, model_layers, batch_size,
+               reg_mf=0):
+    """Initialize NeuMF model.
+
+    Args:
+      num_users: An integer, the number of users.
+      num_items: An integer, the number of items.
+      mf_dim: An integer, the embedding size of Matrix Factorization (MF) model.
+      model_layers: A list of integers for Multi-Layer Perceptron (MLP) layers.
+        Note that the first layer is the concatenation of user and item
+        embeddings. So model_layers[0]//2 is the embedding size for MLP.
+      batch_size: An integer for the batch size.
+      reg_mf: A floating number, the regularization for MF embeddings.
+
+    Raises:
+      ValueError: if the first model layer is not even.
+    """
+    if model_layers[0] % 2 != 0:
+      raise ValueError("The first layer size should be multiple of 2!")
+
+    # Input variables
+    user_input = tf.keras.layers.Input(
+        shape=(1,), dtype="int32", name=constants.USER)
```

Comment:
It's weird that the dtype is a string here. tf.int32 doesn't work?

Reply:
string type is required for keras.layers.Input()